Import BSP project via Fastpass

tpasternak opened the pull request:
This commit implements import of Pants projects to IntelliJ by converting
them to BSP projects with fastpass.

The tricky part was that, contrary to SBT-BSP imports, fastpass creates
BSP configuration *outside* of project root. This means, if the repository
root is `~/pants-example` and one wants to import `~/pants-example/java_app`,
the BSP configuration is not created in `~/pants-example/.bsp`, but in
`~/bsp-projects/java_app/.bsp`.

In order to support this case, the field `externalBspWorkspace` is introduced,
where BSP configuration is generated and stored, even if something else was chosen
by the user to import. Despite that, the BspSetupConfigStep must still receive
the original path selected by the user, because it needs to know what exactly is
supposed to be exported.

Thanks to this behavior, a user can import a few small, independent subprojects simultaneously
without touching rest of the project.

References:
https://github.com/scalameta/fastpass

Commit:
Import BSP project via Fastpass

This commit implements import of Pants projects to IntelliJ by converting
them to BSP projects with fastpass.

The tricky part was that, contrary to SBT-BSP imports, fastpass creates
BSP configuration *outside* of project root. This means, if the repository
root is `~/pants-example` and one wants to import `~/pants-example/java_app`,
the BSP configuration is not created in `~/pants-example/.bsp`, but in
`~/bsp-projects/java_app/.bsp`.

In order to support this case, the field `externalBspWorkspace` is introduced,
where BSP configuration is generated and stored, even if something else was chosen
by the user to import. Despite that, the BspSetupConfigStep must still receive
the original path selected by the user, because it needs to know what exactly is
supposed to be exported.

Thanks to this behavior, a user can import a few small, independent subprojects simultaneously
without touching rest of the project.

References:
https://github.com/scalameta/fastpass

tpasternak commented on bsp/src/org/jetbrains/bsp/project/importing/setup/FastpassConfigSetup.scala:
```diff
@@ -0,0 +1,75 @@
+      }
+    }
+
+    // TODO build messages
```
@jastice I'm not sure how I should handle build messages here

jastice replied:
just giving an OK status is fine, but you could also pass in any collected log messages

jastice commented on bsp/src/org/jetbrains/bsp/settings/settings.scala:
```diff
@@ -11,13 +11,16 @@ import com.intellij.openapi.externalSystem.settings._
+import org.jetbrains.bsp.project.importing.setup.FastpassConfigSetup
 import org.jetbrains.bsp.settings.BspProjectSettings.{AutoConfig, AutoPreImport, BspServerConfig, BspServerConfigConverter, PreImportConfig, PreImportConfigConverter}
 import org.jetbrains.bsp.{BspBundle, _}
 import org.jetbrains.plugins.scala.project.ProjectExt
```
imports unused

tpasternak replied:
done

jastice commented on scala/scala-impl/src/org/jetbrains/sbt/project/FastpassProjectImportProvider.scala:
```diff
@@ -0,0 +1,12 @@
+import com.intellij.openapi.vfs.VirtualFile
+import Iterator.iterate
+
+object FastpassProjectImportProvider {
```
why is this under the sbt package?

tpasternak replied:
ok, moved it. It was there alongside Mill provider

jastice replied:
oh, I thought I had moved that one

jastice commented on bsp/src/org/jetbrains/bsp/project/importing/setup/FastpassConfigSetup.scala:
```diff
@@ -0,0 +1,75 @@
+
+  override def run(implicit reporter: BuildReporter): Try[BuildMessages] = {
+    reporter.start()
+    val process = processBuilder.start()
```
you may want to send process output to the reporter

tpasternak replied:
the messages are now passed to reporter

jastice commented on bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala:
```diff
@@ -14,13 +14,15 @@ import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil
 import org.jetbrains.bsp.{BSP, BspBundle}
 import org.jetbrains.plugins.scala.project.Version
 import org.jetbrains.plugins.scala.project.external.{JdkByHome, JdkByVersion}
+import org.jetbrains.sbt.project.FastpassProjectImportProvider
```
unused imports

tpasternak replied:
done

jastice reviewed (commented):
Thanks! I'll try it out next week!

Commit:
Optimize imports

Commit:
Report `fastpass create` messages

Commit:
Fill BuildMessages in FastpassConfigSetup

Commit:
Optimize imports

jastice commented on bsp/src/org/jetbrains/bsp/project/importing/setup/FastpassConfigSetup.scala:
```diff
@@ -50,18 +50,28 @@ class FastpassConfigSetup(processBuilder: ProcessBuilder) extends BspConfigSetup
+      }
+      stdoutReader.lines().forEach{ line =>
+        buildMessages = buildMessages.message(line)
+        reporter.info(line, None)
```
may want to just use BuildReporter.log - info may be displayed as individual nodes (if build toolwindow is used)

jastice commented:
Do you have a project example that I could try this with?

jastice commented on bsp/src/org/jetbrains/bsp/project/importing/projectImport.scala:
```diff
@@ -203,14 +212,18 @@ class BspProjectImportProvider(builder: BspProjectImportBuilder)
-      new BspSetupConfigStep(context, builder),
+    // Calling super.createSteps here is essential to ensure that
+    // the `NewProjectUtil.doCreate` is called *after* resolveProjectInfo(isPreview=true)
+    super.createSteps(context) ++ Array(
```
This causes an additional dialog on import, which isn't needed most of the time. Can we avoid it?

tpasternak replied:
Ok, managed to do it without the extra step, just commited. Nevertheless, I'm really not sure if it's safe to skip `super.createSteps` here

tpasternak commented:
Ok, it seems this was an issue with my homebrew setup, no with Macs in general. You can try to import `src/java/org/pantsbuild/example/hello/greet` from this repo `https://github.com/scalameta/pants`

tpasternak commented:
BTW - there's an assumption here that there is a `fastpass exec` in your path - you can download it with `coursier bootstrap org.scalameta:fastpass_2.12:1.5.0 -r sonatype:snapshots --main scala.meta.fastpass.Fastpass -o fastpass`

Commit:
Fix NPE

The function is called repeatedly, the basePath is set at some point

jastice commented:
> BTW - there's an assumption here that there is a `fastpass exec` in your path - you can download it with `coursier bootstrap org.scalameta:fastpass_2.12:1.5.0 -r sonatype:snapshots --main scala.meta.fastpass.Fastpass -o fastpass`

That's alright for now, but can we avoid it by bundling coursier? We'll probably want to do this anyway for bazel-bsp.

Commit:
Handle non-existent directory as valid for project import

Commit:
Do not create unnecessary import wizard step

Unfortunately, one of the steps (SelectExternalProjectStep) calls an important function -
builder.ensureProjectIsDefined. After the SelectExternalProjectStep is disabled, it has go
be called manually.

tpasternak commented:
> That's alright for now, but can we avoid it by bundling coursier? We'll probably want to do this anyway for bazel-bsp.

Ideally, I'd prefer it to be somehow configurable, but of course, you're right that bundling coursier would simplify stuff a lot

Commit:
Correctly handle non-existent workspace path

jastice commented:
During the initial setup I'm getting a lot of lines prefixed with `ERROR:` in the indicator

![image](https://user-images.githubusercontent.com/770879/91052084-41c59b00-e621-11ea-8567-287455d4fa1d.png)

tpasternak commented:
> During the initial setup I'm getting a lot of lines prefixed with `ERROR:` in the indicator

This is because I'm calling `reporter.info`and stream all logs from fastpass to to this window. As I said, I didn't know what to do with this log. I haven't tried `reporter.log` yet

jastice commented:
> This is because I'm calling `reporter.info`and stream all logs from fastpass to to this window. As I said, I didn't know what to do with this log. I haven't tried `reporter.log` yet

Yup, looks like you're treating every line from stderr as error, but if that's not how fastpass treats them I'd go with 
```
buildMessages = buildMessages.message(line)
reporter.log(line)
```
for both stderr and stdout

Commit:
Use reporter.log instead of reporter.info

Just for the better user experience

jastice commented:
I imported `/Users/jast/playspace/pants/examples/src/java/org/pantsbuild/example/hello/greet`

After the import is done, I'm getting an empty `.bloop` directory, and correspondingly, this output on refresh:

```
[D] Waiting for a connection at local:///var/folders/5k/4b5my29x26q_dhrys3ghf9wh0000gn/T/bsp-launcher12815398736040296871/bsp.socket...

The server is listening for incoming connections at local:///var/folders/5k/4b5my29x26q_dhrys3ghf9wh0000gn/T/bsp-launcher12815398736040296871/bsp.socket...

Starting thread that pumps stdin and redirects it to the bsp server...

Starting thread that pumps server stdout and redirects it to the client stdout...

Empty build targets. Expected at least one build target identifier.
Empty build targets. Expected at least one build target identifier.
Empty build targets. Expected at least one build target identifier.
```

tpasternak commented:
Oh, could have been fixed in ffab741

jastice commented:
Still getting the same issue

![image](https://user-images.githubusercontent.com/770879/91057897-1c885b00-e628-11ea-86d6-10e72ffa7b02.png)

jastice commented on bsp/src/org/jetbrains/bsp/project/importing/setup/FastpassConfigSetup.scala:
```diff
@@ -55,11 +55,11 @@ class FastpassConfigSetup(processBuilder: ProcessBuilder) extends BspConfigSetup
       val stderrLines = stderrReader.lines()
       stderrLines.forEach{ line =>
         buildMessages = buildMessages.addError(line)
```
use `buildMessages.message` here too if you don't expect `line` to be actual errors in general

Commit:
Correctly interpret subprocess error code

Commit:
Do not run fastpass if BSP configuration already exists

Apart from this, add some logs

tpasternak commented:
@jastice I added support for the case when is BSP project already exists in the path - it's not running fastpass, just imports that BSP project
